graph/patterns/viewpoint.md

```diff
@@ -0,0 +1,112 @@
+    </Annotations>
+```
+- An alternative to this design would be to store the user state on the client side. However, this may be problematic in some cases, because of the many devices that a user may have and the need to synchronize the state across them.
+
+## Examples
+
+### Defining a viewpoint
+
+The following example demonstrates how to define the 'viewpoint' property for the `chat` entity, where a chat is a collection of chatMessages between one or more participants: 
+```
+  <ComplexType Name="chatViewpoint" ags:WorkloadIds="Microsoft.Teams.GraphSvc">
+        <Property Name="isHidden" Type="Edm.Boolean" />
+        <Property Name="lastMessageReadDateTime" Type="Edm.DateTimeOffset" />
+  </ComplexType>
+
+  <EntityType Name="chat" BaseType="graph.entity" ags:OwnerService="Microsoft.Teams.GraphSvc">
+        <Property Name="chatType" Type="graph.chatType" Nullable="false" />
+        <Property Name="createdDateTime" Type="Edm.DateTimeOffset" />
+        <Property Name="lastUpdatedDateTime" Type="Edm.DateTimeOffset" />              
+        <Property Name="topic" Type="Edm.String" />
+        <Property Name="viewpoint" Type="graph.chatViewpoint" />
+       ...
+        <NavigationProperty Name="tabs" Type="Collection(graph.teamsTab)" ContainsTarget="true" />
+  </EntityType>
+
+```
+### Reading an entity with a viewpoint
+
+The following example shows reading a collection of chats for an identified user, with a viewpoint for each chat:
+
+```http
+GET https://graph.microsoft.com/v1.0/users/8b081ef6-4792-4def-b2c9-c363a1bf41d5/chats
+```
+
+```http
+HTTP/1.1 200 OK
+Content-type: application/json
+
+{
+    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#chats",
+    "@odata.count": 3,
+    "value": [
+        {
+            "id": "19:meeting_MjdhNjM4YzUtYzExZi00OTFkLTkzZTAtNTVlNmZmMDhkNGU2@thread.v2",
+            "topic": "Meeting chat sample",
+            "createdDateTime": "2020-12-08T23:53:05.801Z",
+            "lastUpdatedDateTime": "2020-12-08T23:58:32.511Z",
+            "chatType": "meeting",         
+            "viewpoint":{
+                "lastMessageReadDateTime": "2021-03-28T21:10:00.000Z" // User has unread messages
+            }
+        },
+        {
+            "id": "19:561082c0f3f847a58069deb8eb300807@thread.v2",
+            "topic": "Group chat sample",
+            "createdDateTime": "2020-12-03T19:41:07.054Z",
+            "lastUpdatedDateTime": "2020-12-08T23:53:11.012Z",
+            "chatType": "group",            
+            "viewpoint":{
+                "lastMessageReadDateTime": "0000-01-01T00:00:00.000Z" // User hasnt read anything since no message was posted
+            }
+        }
+    ]
+}
+```
+### Updating a viewpoint
+
+You can update the `viewpoint` property only if the server does not compute it automatically. Updating the `viewpoint` property usually has a side effect, so you should use an OData action to perform the update.
```

Comment:
Not sure why this requires an OData action.  Why can't updating the property (via PATCH), initiate whatever other changes need to happen?  I think an action or a PATCH here would be fine, but this reads like the API producer must use an action.

Reply:
Updating a viewpoint requires a change of a shared resource created for saving client-specific state and not exposed via Graph. Therefore, there is an unpredicted side-effect and use of a function. Using PATCH, I'm not sure how to manage concurrency and ETags. IMO action is preferable.

Reply:
I still don't get this.  The viewpoint presented is in the scope of the signed-in user.  So why can't the viewpoint update also be in the context of the signed-in user.

I might buy that ETags could be an interesting challenge here.